std/special/compiler_rt/divdf3.zig

```diff
@@ -0,0 +1,328 @@
+    // is small; most of the work is already done.
+    reciprocal -%= 2;
+
+    // The numerical reciprocal is accurate to within 2^-56, lies in the
+    // interval [0.5, 1.0), and is strictly smaller than the true reciprocal
+    // of b.  Multiplying a by this reciprocal thus gives a numerical q = a/b
+    // in Q53 with the following properties:
+    //
+    //    1. q < a/b
+    //    2. q is in the interval [0.5, 2.0)
+    //    3. the error in q is bounded away from 2^-53 (actually, we have a
+    //       couple of bits to spare, but this is all we need).
+
+    // We need a 64 x 64 multiply high to compute q, which isn't a basic
+    // operation in C, so we need to be a little bit fussy.
+    var quotient: Z = undefined;
+    var quotientLo: Z = undefined;
+    wideMultiply(Z, aSignificand << 2, reciprocal, &quotient, &quotientLo);
+
+    // Two cases: quotient is in [0.5, 1.0) or quotient is in [1.0, 2.0).
+    // In either case, we are going to compute a residual of the form
+    //
+    //     r = a - q*b
+    //
+    // We know from the construction of q that r satisfies:
+    //
+    //     0 <= r < ulp(q)*b
+    //
+    // if r is greater than 1/2 ulp(q)*b, then q rounds up.  Otherwise, we
+    // already have the correct result.  The exact halfway case cannot occur.
+    // We also take this time to right shift quotient if it falls in the [1,2)
+    // range and adjust the exponent accordingly.
+    var residual: Z = undefined;
+    if (quotient < (implicitBit << 1)) {
+        residual = (aSignificand << 53) -% quotient *% bSignificand;
+        quotientExponent -%= 1;
+    } else {
+        quotient >>= 1;
+        residual = (aSignificand << 52) -% quotient *% bSignificand;
+    }
+
+    const writtenExponent = quotientExponent +% exponentBias;
+
+    if (writtenExponent >= maxExponent) {
+        // If we have overflowed the exponent, return infinity.
+        return @bitCast(f64, infRep | quotientSign);
+    } else if (writtenExponent < 1) {
+        if (writtenExponent == 0) {
+            // Check whether the rounded result is normal.
+            const round = @boolToInt((residual << 1) > bSignificand);
+            // Clear the implicit bit.
+            var absResult = quotient & significandMask;
+            // Round.
+            absResult += round;
+            if ((absResult & ~significandMask) != 0) {
+                // The rounded result is normal; return it.
+                return @bitCast(f64, absResult | quotientSign);
+            }
+        }
+        // Flush denormals to zero.  In the future, it would be nice to add
+        // code to round them correctly.
+        return @bitCast(f64, quotientSign);
+    } else {
+        const round = @boolToInt((residual << 1) > bSignificand);
+        // Clear the implicit bit
+        var absResult = quotient & significandMask;
+        // Insert the exponent
+        absResult |= @bitCast(Z, SignedZ(writtenExponent)) << significandBits;
```

Comment:
This line was the only thing I had to look up to make sure I understood.
It cast `writtenExponent` to a `rep_t` which is a cast from `i32` to `u64`. C keeps the bit pattern of the same numerical value but at the new type width when doing a cast like this. ie -1 as an i32 becomes 0xFFFFFFFF FFFFFFFF as a u64.

Reply:
Also just trying out the comment system :)